administrator/components/com_fields/helpers/fields.php

```diff
@@ -161,6 +161,9 @@ public static function getFields ($context, $item = null, $prepareValue = false)
 				}
 
 				$field->rawvalue = $field->value;
+				
+				//Store item id on field instance for later use			
+				$field->item_id = $item->id;
```

Comment:
What about removing the item it assignment here and adding a the item as parameter to the trigger. Something like 

    $dispatcher->trigger('onFieldBeforePrepare',array($context, $item, &$field));

and then on also

    $dispatcher->trigger('onFieldAfterPrepare',array($context, $item, $field, &$value));